// drizzle-orm/supabase/rls.js
import { bigserial, pgSchema, text, uuid } from "../pg-core/index.js";
import { pgRole } from "../pg-core/roles.js";
import { sql } from "../sql/sql.js";
const anonRole = pgRole("anon").existing();
const authenticatedRole = pgRole("authenticated").existing();
const serviceRole = pgRole("service_role").existing();
const postgresRole = pgRole("postgres_role").existing();
const supabaseAuthAdminRole = pgRole("supabase_auth_admin").existing();
const auth = pgSchema("auth");
const authUsers = auth.table("users", {
  id: uuid().primaryKey().notNull()
});
const realtime = pgSchema("realtime");
const realtimeMessages = realtime.table(
  "messages",
  {
    id: bigserial({ mode: "bigint" }).primaryKey(),
    topic: text().notNull(),
    extension: text({
      enum: ["presence", "broadcast", "postgres_changes"]
    }).notNull()
  }
);
const authUid = sql`(select auth.uid())`;
const realtimeTopic = sql`realtime.topic()`;
export {
  anonRole,
  authUid,
  authUsers,
  authenticatedRole,
  postgresRole,
  realtimeMessages,
  realtimeTopic,
  serviceRole,
  supabaseAuthAdminRole
};
